drivers/flash/flash_ambiq.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2023 Ambiq Micro Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ambiq_flash_controller
+
+#include <zephyr/kernel.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+
+#include <am_mcu_apollo.h>
+
+LOG_MODULE_REGISTER(flash_ambiq, CONFIG_FLASH_LOG_LEVEL);
+
+#define SOC_NV_FLASH_NODE      DT_INST(0, soc_nv_flash)
+#define SOC_NV_FLASH_ADDR      DT_REG_ADDR(SOC_NV_FLASH_NODE)
+#define SOC_NV_FLASH_SIZE      DT_REG_SIZE(SOC_NV_FLASH_NODE)
+#define FLASH_WRITE_BLOCK_SIZE DT_PROP(SOC_NV_FLASH_NODE, write_block_size)
+#define FLASH_ERASE_BLOCK_SIZE DT_PROP(SOC_NV_FLASH_NODE, erase_block_size)
+
+struct flash_ambiq_data {
+#ifdef CONFIG_MULTITHREADING
+	struct k_sem mutex;
+#endif /* CONFIG_MULTITHREADING */
+};
+
+static struct flash_ambiq_data flash_data;
+
+static const struct flash_parameters flash_ambiq_parameters = {
+	.write_block_size = FLASH_WRITE_BLOCK_SIZE,
+	.erase_value = 0xff,
+};
+
+static bool flash_ambiq_valid_range(off_t offset, size_t len, bool read)
+{
+	if ((offset < 0) || (offset + len) > SOC_NV_FLASH_SIZE) {
+		return false;
+	}
+
+	if (!read) {
+		/* The address and length of write and erase need to be 16 bytes aligned.*/
+		if (((uint32_t)offset & 0xf) || (len & 0xf)) {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static int flash_ambiq_read(const struct device *dev, off_t offset, void *data, size_t len)
+{
+	ARG_UNUSED(dev);
+
+	if (!flash_ambiq_valid_range(offset, len, true)) {
+		return -EINVAL;
+	}
+
+	if (len == 0) {
+		return 0;
+	}
+
+	memcpy(data, (uint8_t *)(SOC_NV_FLASH_ADDR + offset), len);
+
+	return 0;
+}
+
+static int flash_ambiq_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	struct flash_ambiq_data *dev_data = dev->data;
+	int ret = 0;
+	uint32_t critical = 0;
+
+	if (!flash_ambiq_valid_range(offset, len, false)) {
+		return -EINVAL;
+	}
+
+	if (len == 0) {
+		return 0;
+	}
+
+#ifdef CONFIG_MULTITHREADING
+	k_sem_take(&dev_data->mutex, K_FOREVER);
+#endif /* CONFIG_MULTITHREADING */
+
+	critical = am_hal_interrupt_master_disable();
+	ret = am_hal_mram_main_program(AM_HAL_MRAM_PROGRAM_KEY, (uint32_t *)data,
+				       (uint32_t *)(SOC_NV_FLASH_ADDR + offset), (len / 4));
+	am_hal_interrupt_master_set(critical);
+
+#ifdef CONFIG_MULTITHREADING
+	k_sem_give(&dev_data->mutex);
+#endif /* CONFIG_MULTITHREADING */
+
+	return ret;
+}
+
+static int flash_ambiq_erase(const struct device *dev, off_t offset, size_t len)
```

Comment:
MRAM does not require erase and does not provide it.
Emulated erase, in case of any xRAM devices, reduces use count by 1 the same as write does. In case of flash, the reduction only happens at erase operation, write does not affect wear.